docker/docker_image_dest.go

```diff
@@ -351,8 +371,24 @@ func (d *dockerImageDestination) TryReusingBlob(ctx context.Context, info types.
 				continue
 			}
 		}
-		cache.RecordKnownLocation(d.ref.Transport(), bicTransportScope(d.ref), candidate.Digest, newBICLocationReference(d.ref))
-		return true, types.BlobInfo{Digest: candidate.Digest, Size: size}, nil
+
+		bic.RecordKnownLocation2(d.ref.Transport(), bicTransportScope(d.ref), candidate.Digest, candidate.CompressorName, newBICLocationReference(d.ref))
+
+		var compressionOperation types.LayerCompression
+		var compressionAlgorithm *compression.Algorithm
+		if candidate.CompressorName != blobinfocache.Uncompressed {
+			algorithm, err := compression.AlgorithmByName(candidate.CompressorName)
+			if err != nil {
+				logrus.Debugf("... unknown compressor %q", candidate.CompressorName)
+				return false, types.BlobInfo{}, nil
+			}
+			compressionOperation = types.Compress
+			compressionAlgorithm = &algorithm
+		} else {
+			compressionOperation = types.Decompress
+			compressionAlgorithm = &compression.Algorithm{}
+		}
```

Comment:
Nothing about this is transport-specific; it feels like something that should exist in the generic code.

Reply:
Moved to a helper function in `internal/blobinfocache` for use by `TryReusingBlob()` implementations.